Restructure Ansible scripts

alexling92 opened the pull request.

alexling92 committed:
restructure ansible scripts

alexling92 committed:
Update READEME.md

eugene-chow commented on Monitoring in a Box/Ansible/group_vars/all.yml:
```diff
@@ -60,4 +31,12 @@ DNSIP: 10.0.1.53
-SshJumpHost: ProxyJump [key in the public ip and port number]
+SshJumpHost: ProxyJump 111.223.106.169:22222
+
+DBpass: !vault |
+          $ANSIBLE_VAULT;1.1;AES256
+          32393339373438313935656365646337316466386635316164323533656138323230343862373032
+          3062373738363362393532626262343237333137346534350a616461366665653462383630313665
+          33613936663864343834643838376365356633643362623534326264306537373165623539616363
+          6633393132623537380a363863333333323761373232396438323530666133396366383538643436
+          3463
```
The database password should be a hostvar. Each deployment should have its own password.

alexling92 replied:
I was thinking of putting it into host_var at first, but on the WebGui side, the password is hard-coded into the python scripts. Therefore it would not be change, thats why i decided to put at group_vars. This way the user wont accidentally edit it also.

eugene-chow replied:
Don't hard-code the password into the python script. What needs to be done is to put this script into the role's `template` folder and then use Ansible's `template` module to paste in the password.

alexling92 replied:
I will discuss with john to edit it accordingly

alexling92 replied:
Resolved in `Remove hardcoded DB password` commit.

eugene-chow commented on Monitoring in a Box/Ansible/group_vars/all.yml:
```diff
@@ -30,26 +17,10 @@ CreateFullClone: --full true
-  - 10.0.1.240
-
 ### Network gateway ###
 GwIP: 10.0.1.5
```
Gateway IP and DNS search domain should be a hostvar. Every environment is different. Actually DNS search domain is unnecessary so you can omit it.

alexling92 replied:
Because this script is specific to our environment , for example the pve node to create on and also the image template name. Which is why I decided to put the gateway IP in the group_vars , since its already dependent on the environment and should not be change

eugene-chow replied:
Environment-related vars is specific to the environment so it should not be a global var in `all.yml`. The rationale for putting it in hostvars is to have 1x hostvar per deployment.

alexling92 replied:
This gateway ip is actually part of the set up environment in the test cluster, so no matter which host, they will be using this gateway ip for internet access during the setup. Also I worried that if i put it in the host_var, the user might accidentally edit it. Which is why I put it in all.yml, out of sight from the user.

eugene-chow replied:
I think there's a misunderstanding. The Ansible scripts will not be run by the user so only you guys should be using it. Is there something else I'm missing?

alexling92 replied:
Yeah this is so we can create and set up a MiB VM. After which we will just export the VM image and give the user. Then we will advice user to access the /etc/netplan config to set up the network to suit their environment.

eugene-chow replied:
Correct lor. That means the vars can be in hostvars.

alexling92 replied:
is it? even tho all the VMs being set up will use the same gateway ip? 
Just to clarify any misunderstanding, here are some pointers to note:
1. This gateway ip is to assist us in the creation and setting up of the MiB VM in our test cluster environment. i.e to provide internet access to install the dependencies.
1. The User will not be using this ansible script at all.
1. The User will only receive the VM image from us.

eugene-chow replied:
Ok sorry I understand your point now. Rename the variables to better describe what's going on and comment the code as well. That way there's lesser confusion. Thanks!

eugene-chow commented on Monitoring in a Box/Ansible/ansible.cfg:
```diff
@@ -0,0 +1,3 @@
+[defaults]
+host_key_checking = False
+vault_password_file = .vault
```
Remember to put the vault password in 1Password and share with the team.

alexling92 replied:
okay

eugene-chow commented on Monitoring in a Box/Ansible/inventory.yaml:
```diff
@@ -1,17 +1,5 @@
-
 [mib_host]
-mib@10.0.1.240
+mib1  ansible_user=mib	ansible_host=10.0.1.240
```
Can this be part of the hostvar? Have a single source of truth as far as possible.

alexling92 replied:
I will shift it to host_var and test if it works

alexling92 replied:
I tested it already and it works. I will be using the following set up to link the variables together:
ansible_user: "{{NewVmUsername}}"
ansible_host: "{{NewVmIP}}"

eugene-chow replied:
Cool 👍

alexling92 replied:
I will also be shifting the ssh jumphost argument into the host_vars file to consolidate all the variables which require editing in a single file. Therefore user only needs to access host_vars to fill in all the necessary variables.

alexling92 replied:
I have shifted the variables in the `Shift Inventory variables to host_vars file` commit

eugene-chow commented on Monitoring in a Box/Ansible/inventory.yaml:
```diff
@@ -1,17 +1,5 @@
 
 [mib_host:vars]
-ansible_become_pass=[key in the password set for mib vm]
+ansible_ssh_common_args='-o ProxyCommand="ssh -W %h:%p -q 111.223.106.169 -p 22222"'
```
Is this a duplicate of the ProxyJump command in `all.yml`?

alexling92 replied:
My bad I forgot to remove the `sshjumphost` variable in all.yml. This is a replacement for the setting up of ssh jumphost task.

alexling92 replied:
I have removed it in `Remove SshJumpHost variable` commit.

eugene-chow reviewed (changes requested):
Cultivate the habit of creating a commit per change. Put yourself in the shoes of the reviewer before you make a PR. @jiunnyuan1991 please advise Alex and the rest how to make multiple commits.

I'll reject such PRs by default in the future. @youngee91 @alexchoo123 take note.

alexling92 commented:
> Cultivate the habit of creating a commit per change. Put yourself in the shoes of the reviewer before you make a PR. @jiunnyuan1991 please advise Alex and the rest how to make multiple commits.
> 
> I'll reject such PRs by default in the future. @youngee91 @alexchoo123 take note.

My bad for submitting all the changes in 1 commit, I was working on it in another directory to test and make sure it works. After which I just copy paste the whole ansible directory, next time I will make sure to split up changes in to different commit so that it will be more organise for the reviewer.

eugene-chow commented:
Thanks. It's ok to make mistakes the first time. Just don't repeat them.

alexling92 committed:
Remove SshJumpHost variable

alexling92 committed:
Shift Inventory variables to host_vars file

I have also shifted the jumphost argument into the host_vars file, therefore user will only need to access host_vars file to edit the necessary variables.

eugene-chow commented on Monitoring in a Box/Ansible/host_vars/mib1.yaml:
```diff
@@ -13,11 +14,18 @@ NewVmName: "MiB-Host"
+# SSH VM IP address
+ansible_host: "{{NewVmIP}}"
+# SSH jumphost option
+ansible_ssh_common_args: '-o ProxyCommand="ssh -W %h:%p -q [Key in Jumphost IP] -p [Key in Jumphost Port Number]"'
```
If the Ansible script will only be used in your environment, shouldn't it be in `all.yml`? I am confused because the gateway IP and DNS are in `all.yml` because this is for your environment. `ansible_ssh_common_args` should rightfully also be for your environment.

alexling92 replied:
Yeah that should be the case, however I did not want to expose the public IP address and also decided to consolidate all the variables which requires editing into a single file to make it easier for the whole setting up process .

eugene-chow replied:
I understand your train of thought but it's not the right way. When writing code, assume that someone one day will take charge of your code. The poor fella will be confused.

Since this is a global variable for your environment by your definition, it should end up in `all.yml` like the rest. If every single hostvar will use the same `ansible_ssh_common_args`, then it should be treated as a common variable and put either in a groupvar in `all.yml` or `mib_host.yml`.

If you don't wish to expose the public IP, use Ansible Vault to wrap it up as an encrypted variable.

alexling92 replied:
Roger. I will make the necessary changes and also do the same for similar variables.

eugene-chow replied:
Cool thanks!

alexling92 committed:
Move variable from host_vars to group_vars

alexling92 committed:
Update README.md

alexling92 committed:
Remove hardcoded DB password

alexling92 committed:
Create .gitignore

alexling92 committed:
Remove all .pyc file

eugene-chow commented on Monitoring in a Box/Ansible/host_vars/mib1.yaml:
```diff
@@ -0,0 +1,30 @@
+NewVmIP: 10.0.1.240
+
+# Pve node sudo password
+ansible_become_pass: [key in pve node sudo password]
```
Sorry can you remind me what this is? I believe the `--ask-become-pass` flag will provide the required password.

alexling92 replied:
This is the same as the `--ask-become-pass` flag use with the ansible command. Except that you specify it in the vars or inventory folder to avoid the password prompt and keying the password.

eugene-chow replied:
In this case, omit the variable because putting the password here is an operational risk. Raw passwords should never be included in the Ansible script if avoidable. If you commit the code with the password, you'll have to reset the password on all affected nodes.

The more sound way is to specify `--ask-become-pass`.

alexling92 replied:
Alright I will remove the `ansible_become_pass` variable from the script.

eugene-chow commented on Monitoring in a Box/Ansible/host_vars/mib1.yaml:
```diff
@@ -0,0 +1,30 @@
+# Pve node sudo password
+ansible_become_pass: [key in pve node sudo password]
+
+## New account password must be the same as pve node sudo password
+NewVmPassword: "{{ansible_become_pass}}"
```
Actually not true. The Ubuntu template's default setting enables passwordless sudo. This means the Ansible script can configure the VM. After the VM is configured, set the password for the `mib` user as the very last task.

alexling92 replied:
Then can I set a default password on the MiB VM, and ask the user to change it when they receive the vm?

alexling92 replied:
I tried it out with different password from the pve node sudo password but the script will fail with this error `fatal: [mib1]: FAILED! => {"msg": "Incorrect sudo password"}`

alexling92 replied:
I have to shift the `ansible_become_pass` variable for pve-node to `roles/create_vm/vars/main.yaml` file for the script to work. The variable is shifted there because for the creation of vm, need to become sudo user for the pve node.

eugene-chow replied:
Like the DB credentials, a unique password should be set for each VM. The user should be advise to reset the password after receiving the appliance (add this to the documentation).

> I have to shift the ansible_become_pass variable for pve-node to roles/create_vm/vars/main.yaml file for the script to work. The variable is shifted there because for the creation of vm, need to become sudo user for the pve node.

Can you check if the VM's default behaviour permits passwordless sudo? I believe that's the case. If it's not, then what you could do is to first set the `mib` user password (via cloud-init) to `ansible_become_pass`. After the VM is fully configured, set the password to the deployment-specific one.

alexling92 replied:
> Like the DB credentials, a unique password should be set for each VM. The user should be advise to reset the password after receiving the appliance (add this to the documentation).

Ok, I will use ansible vault to set a common password for all MiB VM. 

> Can you check if the VM's default behaviour permits passwordless sudo? I believe that's the case. If it's not, then what you could do is to first set the mib user password (via cloud-init) to ansible_become_pass. After the VM is fully configured, set the password to the deployment-specific one.

Yup the VM, default behaviour is passwordless sudo. I did a test with the `ask-become-pass` flag and it is able to deploy successfully without any issue. I will remove the `ansible_become_pass` variable.

alexling92 replied:
1. I do have a question, if we are going to set the default user as "mib" and also the default password, should we put it in the `group_vars/all.yml` file instead? Since it will be the same for all the MiB VMs being created and set up(before delivery to the end user).

1. Since we will be setting a default password and advising the user to change the password upon receiving the MiB VM, do we have the use ansible vault to encrypt it? (Because those major appliances in the market, their default password can be found easily online)

eugene-chow replied:
1. Ideally, the `mib` user in each VM should be configured with a unique password. This can be specified as a vaulted variable in the deployment's hostvar. There are 2 ways to handle deployments:
   1. Set a default password for all VMs. Make the client reset the password as part of the SOP. Add this to the documentation.
   1. Set a unique password for each deployment and leave it to the client to reset the password. I personally think this is safer but I leave it up to you.
1. Always put passwords, secrets and sensitive info in a vault. Always practice good security hygiene by default. "Major appliances" such as home routers are in a different category because they can be reset to a default password if the config is messed up.

alexling92 replied:
> 1. Ideally, the `mib` user in each VM should be configured with a unique password. This can be specified as a vaulted variable in the deployment's hostvar. There are 2 ways to handle deployments:
>    
>    1. Set a default password for all VMs. Make the client reset the password as part of the SOP. Add this to the documentation.
>    2. Set a unique password for each deployment and leave it to the client to reset the password. I personally think this is safer but I leave it up to you.

I will go with the first approach because the 2nd approach will be quite tedious during setup as we will need to constantly generate a new unique vault password. I will set the MiB VM password to expire using `passwd -e mib` so that it will force the user to key in a new password on the next login

> 2. Always put passwords, secrets and sensitive info in a vault. Always practice good security hygiene by default. "Major appliances" such as home routers are in a different category because they can be reset to a default password if the config is messed up.

Got it! Thanks

eugene-chow replied:
Ok but the deployment SOP must include telling the user to reset the password. If not an attacker at some point can reset the password.

alexling92 replied:
I will just expire the password, so user definitely has to set a new password at first login.

eugene-chow commented:
Good work so far. Just work on the last 2 things and this PR can be merged. Keep in mind the lessons on security and Ansible best practices.

alexling92 committed:
Remove `ansible_become_pass` variable

alexling92 committed:
Add `set_expire_password` task

alexling92 committed:
create `add_end_user_sshpubkey` task

Remove all ssh public key required during the MiB VM setup phase and add end user's public key to the VM

eugene-chow commented on Monitoring in a Box/Ansible/roles/mib_setup/tasks/set_expire_password.yaml:
```diff
@@ -0,0 +1,2 @@
+- name: set password to expired
+  shell: passwd -e "{{NewVmUsername}}"
```
Use the `user` module with the `expires` parameter.

alexling92 replied:
Did try to use `user` module, but the `expires` parameter is for user expiration (the user account, not the password)

eugene-chow replied:
I see. Good find.

eugene-chow commented on Monitoring in a Box/Ansible/roles/mib_setup/tasks/main.yaml:
```diff
@@ -22,5 +22,8 @@
 - name: Set up MiB service
   import_tasks: set_up_mib_service.yaml
 
+- name: Remove SSH authorized keys from MiB VM
```
It should be "Add" and not "Remove" right?

alexling92 replied:
This task was previously created to remove all ssh pub keys in the VM environment and I forgot to change the name. Will edit it to "add end user ssh public key" instead

alexling92 committed:
Edit task name

eugene-chow reviewed (approved):
Very good work. Thanks!
